### PR TITLE
JS-2352 Implement rule S7503: Async functions should use async features

### PR DESCRIPTION
**Pull request by martin-strecker-sonarsource**
Reuses the existing S7503 (currently Python-only, "Async functions should use async features") for JS/TS instead of minting a new rule number — it's an exact semantic match for typescript-eslint's `require-await`. Companion RSPEC PR to extend S7503 to JS/TS is tracked separately and not yet merged; the `S7503.json` resource in this PR is a hand-authored stub mirroring the intended RSPEC metadata and will be overwritten by `rspec:refresh` once that PR lands.

## Summary
- Decorates `@typescript-eslint/require-await` as a plain metadata wrapper (`packages/analysis/src/jsts/rules/S7503/`) — no custom FP-suppression logic.
- Investigated every community-reported "false positive" against `require-await` on the typescript-eslint tracker before deciding not to add exceptions. All were either already fixed upstream, deliberate maintainer-defended design, or reporter misunderstandings:
  - #11669 (async generators/iterators) — reporter agreed it wasn't a bug, closed it themselves.
  - #11425 (`await using` support) — closed as duplicate; already implemented upstream (#7866).
  - #11819 (`AsyncDisposableStack`) — maintainer confirmed the callback type accepts sync functions fine; report was simply wrong.
  - #11973 / #11085 (interface/decorator contracts requiring async) — maintainer: using `async` purely to coerce a return type into a `Promise` is "exactly what the rule exists to prevent"; rejected on principle.
  - #10368 (explicit `Promise<T>` return type) — maintainer rejected the proposal outright; the suggested workaround wouldn't even type-check.
  - #11731 (rule shouldn't be recommended by default) — a policy objection, not a detection bug; closed `wontfix` for lack of community consensus.
  - #11093 — a false *negative* in the other direction, unrelated to any of the above.
- Added `S7503.json` (Sonar-side RSPEC resource stub), `meta.ts`, `decorator.ts`, `index.ts`, `unit.test.ts`, and `fixtures/` (tsconfig targeting `esnext` for async generator / `await using` syntax).

## Test plan
- [x] `npx tsx --test packages/analysis/src/jsts/rules/S7503/unit.test.ts` passes (valid cases cover await, empty async functions, async generators, `await using`; invalid cases cover missing await and the interface/decorator-contract pattern rejected above).
- [ ] CI green
- [ ] Companion RSPEC PR merged and `rspec:refresh` re-run to replace the hand-authored `S7503.json` stub

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: sonar-plugin/javascript-checks/src/main/resources/org/sonar/l10n/javascript/rules/javascript/S7503.json [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+    "func": "Constant/Issue",
+    "constantCost": "5min"
+  },
+  "tags": [
+    "async",
+    "type-dependent"
+  ],
+  "defaultSeverity": "Minor",
+  "ruleSpecification": "RSPEC-7503",
+  "sqKey": "S7503",
```
[the remark this answers is not in the document]
Comment (martin-strecker-sonarsource) [reply sonar-plugin/javascript-checks/src/main/resources/org/sonar/l10n/javascript/rules/javascript/S7503.json:15]: Fixed by https://github.com/SonarSource/rspec/pull/8065

File: packages/analysis/src/jsts/rules/S7503/unit.test.ts [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+ * along with this program; if not, see https://sonarsource.com/license/ssal/
+ */
+import { rule } from './index.js';
+import { RuleTester } from '../../../../tests/jsts/tools/testers/rule-tester.js';
+import { describe, it } from 'node:test';
+
+import path from 'node:path';
+import parser from '@typescript-eslint/parser';
+
+describe('S7503', () => {
```
Comment (guillemsarda) [inline packages/analysis/src/jsts/rules/S7503/unit.test.ts:24]: A few edge cases from the investigation aren't locked in by tests yet, worth adding as regression tests so they don't get "fixed" again later:

- **Arrow function with expression body returning a thenable** (no block, no `await`) — the wrapped rule has a dedicated "obviously correct" branch for this, untested here:
  ```ts
  const f = async () => bar(); // valid, no missing-await warning
  declare function bar(): Promise<number>;
  ```
- **Async generator that doesn't yield a thenable** — only the valid (exempt) generator case is tested; add the invalid counterpart to prove the exemption is narrow:
  ```ts
  async function* g() {
    yield 1; // invalid, no await/thenable yield
  }
  ```
- **`AsyncDisposableStack` (#11819)** and **explicit `Promise<T>` return type (#10368)** — both investigated and rejected in the description, but neither has a test proving they're still flagged. Suggest adding an invalid case for each.
- **`for await (...)`** — separate code path (`ForOfStatement[await=true]`) from `AwaitExpression` for marking "has await"; untested:
  ```ts
  async function f() {
    for await (const x of gen()) {}
  }
  declare function gen(): AsyncIterable<number>;
  ```

**Review by guillemsarda — APPROVED** [review]
Left a comment on some examples worth adding as known valid cases.

**Comment by martin-strecker-sonarsource** [comment]
@guillemsarda **Ruling results for the new `S7503` rule — 373 issues across 9 projects**

Ran the ruling baseline update for the 9 projects that surfaced new findings (`ant-design`, `courselit`, `desktop`, `eigen`, `fresh`, `moose`, `searchkit`, `vitest`, `vuetify`). Wanted to flag the volume before merging, since it's higher than most new-rule PRs:

| Project | Issues | Files |
|---|---|---|
| eigen | 169 | 86 |
| desktop | 108 | 56 |
| vitest | 58 | 40 |
| fresh | 14 | 9 |
| ant-design | 12 | 7 |
| courselit | 5 | 5 |
| moose | 5 | 5 |
| searchkit | 1 | 1 |
| vuetify | 1 | 1 |
| **Total** | **373** | |

**Composition:** ~61% (227) land in test files, ~39% (146) in production code — the latter mostly concentrated in a handful of files with repeated patterns (e.g. `desktop`'s `error-handlers.ts`/`app-store.ts` have a family of `async`-typed handler methods with no internal `await`, 11 and 13 hits respectively from the same architectural shape, not independent issues).

**Verified as TPs**, not classification noise: spot-checked ~15 findings across projects/buckets — every one is a genuine `async` function/method with no `await` anywhere in its body, matching zero of the documented exceptions.

One nuance worth knowing: a chunk of the "test file" hits (`vitest`, `ant-design`, `fresh`, `eigen`) are actually a byproduct of the ruling projects' test-file classification gaps (missing/incomplete `testDir` config, or naming conventions — Deno's `_test.ts`, or `eigen`'s plural `.tests.tsx` — that our `isTestRelatedFile` heuristic doesn't recognize), not something specific to S7503. Any Main-scoped rule hits the same files. Not a defect in this PR, just context for why the count looks large.

Given `defaultQualityProfiles: ["Sonar way"]` is already set per your call on the RSPEC PR — are you comfortable with this volume for a Sonar Way rule, or would you rather reconsider before merging?
